api/query_engine.go

```diff
@@ -61,47 +62,60 @@ var (
 //    For "fairness" across series, and because we used to simply reduce any series without regard for how it would be used, we pick the latter. better would be both
 // 3) subject to max-points-per-req-hard: reject the query if it can't be met
 //
-// note: it is assumed that all requests have the same from & to.
 // also takes a "now" value which we compare the TTL against
 
 // TODO: MDP-yes and max-points-per-req-soft code paths may not take into account that archive 0 may have a different raw interval.
 // see https://github.com/grafana/metrictank/issues/1679 (for MDP-no it does do the right thing)
-func planRequests(now, from, to uint32, reqs *ReqMap, planMDP uint32, mpprSoft, mpprHard int) (*ReqsPlan, error) {
+func planRequests(now uint32, reqs *ReqMap, planMDP uint32, mpprSoft, mpprHard int) (*ReqsPlan, error) {
 
 	ok, rp := false, NewReqsPlan(*reqs)
 
 	// 1) Initial parameters
+	getTimeWindowSuperSet := func(reqs ReqsByRet) (uint32, uint32) {
+		minFrom := uint32(math.MaxUint32)
+		maxTo := uint32(0)
+		for _, rs := range reqs {
+			for _, r := range rs {
+				minFrom = util.Min(minFrom, r.From)
+				maxTo = util.Max(maxTo, r.To)
+			}
+		}
+		return minFrom, maxTo
+	}
 	for group, split := range rp.pngroups {
+		// Find the minFrom and maxTo of reqs in the same split
 		if split.mdpyes.HasData() {
-			ok = planLowestResForMDPMulti(now, from, to, planMDP, split.mdpyes)
+			groupFrom, groupTo := getTimeWindowSuperSet(split.mdpyes)
+			ok = planLowestResForMDPMulti(now, groupFrom, groupTo, planMDP, split.mdpyes)
 			if !ok {
 				return nil, errUnSatisfiable
 			}
 			rp.pngroups[group] = split
 		}
 		if split.mdpno.HasData() {
-			ok = planHighestResMulti(now, from, to, split.mdpno)
+			groupFrom, groupTo := getTimeWindowSuperSet(split.mdpno)
+			ok = planHighestResMulti(now, groupFrom, groupTo, split.mdpno)
 			if !ok {
 				return nil, errUnSatisfiable
 			}
 		}
 	}
 	for schemaID, reqs := range rp.single.mdpyes {
-		if len(reqs) == 0 {
-			continue
-		}
-		ok = planLowestResForMDPSingles(now, from, to, planMDP, uint16(schemaID), reqs)
-		if !ok {
-			return nil, errUnSatisfiable
+		// Singles should be planned independently
```

Comment:
I don't have high hopes for MDP-optimization becoming stable any time soon. perhaps we should remove that code.